src/modulus_convert.rs

```diff
@@ -0,0 +1,548 @@
+        shares_of_r1r2: &[ReplicatedSecretSharing],
+        d2_values: &[u64],
+        p: u64,
+        batch_size: usize,
+    ) -> (Vec<ReplicatedSecretSharing>, Vec<u64>) {
+        // To send to helper 1
+        let mut d3_values = Vec::with_capacity(batch_size);
+        let mut output_shares = Vec::with_capacity(batch_size);
+        for i in 0..(batch_size - 1) {
+            let alpha_1 = sub_mod_p(
+                u64::from(r_binary_shares[i].sh2),
+                mult_mod_p(2, shares_of_r1r2[i].sh2, p),
+                p,
+            );
+            let s_2_3: u64 = rng_2_3.gen_range(0..p);
+            let r3: u64 = u64::from(r_binary_shares[i].sh1);
+            let d_3 = sub_mod_p(alpha_1 * r3, s_2_3, p);
+            d3_values.push(d_3);
+            output_shares.push(ReplicatedSecretSharing {
+                sh1: sub_mod_p(r3, mult_mod_p(2, add_mod_p(d2_values[i], s_2_3, p), p), p),
+                sh2: sub_mod_p(alpha_1, mult_mod_p(2, d_3, p), p),
+            });
+        }
+        (output_shares, d3_values)
+    }
+
+    // a and b should both be less than p
+    #[must_use]
+    pub fn add_mod_p(a: u64, b: u64, p: u64) -> u64 {
+        let offset: u64 = u64::MAX - p + 1;
+
+        let (sum, overlow) = a.overflowing_add(b);
+        if overlow {
+            return sum + offset;
+        } else if sum >= p {
+            return sum - p;
+        }
+        sum
+    }
+
+    // a and b should both be less than p
+    #[must_use]
+    pub fn sub_mod_p(a: u64, b: u64, p: u64) -> u64 {
+        let offset: u64 = u64::MAX - p + 1;
+
+        let (difference, overlow) = a.overflowing_sub(b);
+        if overlow {
+            return difference - offset;
+        }
+        difference
+    }
+
+    // a and b should both be less than p
+    #[must_use]
+    pub fn mult_mod_p(a: u64, b: u64, p: u64) -> u64 {
+        let (a, b, p) = (u128::from(a), u128::from(b), u128::from(p));
+
+        // There will not be any trunction, because "mod p" returns
+        // a value less than p, and p is less than u64
+        #[allow(clippy::cast_possible_truncation)]
+        (((a * b) % p) as u64)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use rand::rngs::StdRng;
+    use rand::SeedableRng;
+
+    use crate::modulus_convert::modulus_conversion::add_mod_p;
+    use crate::modulus_convert::modulus_conversion::gen_r_binary_sharings;
+    use crate::modulus_convert::modulus_conversion::mult_mod_p;
+
+    use crate::modulus_convert::modulus_conversion::mult_r1_r2_helper1;
+    use crate::modulus_convert::modulus_conversion::mult_r1_r2_helper2;
+    use crate::modulus_convert::modulus_conversion::mult_r1_r2_helper3;
+
+    use crate::modulus_convert::modulus_conversion::xor_r1_r2_r3_helper1;
+    use crate::modulus_convert::modulus_conversion::xor_r1_r2_r3_helper2;
+    use crate::modulus_convert::modulus_conversion::xor_r1_r2_r3_helper3;
+
+    use crate::modulus_convert::modulus_conversion::sub_mod_p;
```

Comment:
I assume there is a much better way to do this =). What's the Rust best practice for this? Should I put all of these methods on some kind of object and then import that object? If so, what type of object should I use?

Reply:
A lot of those effectively operate on bignums.  There's no reason for us to go beyond u64 (or much beyond it) for the base value.

I'm concerned at ff not providing constant-time code.  That is probably something we need here, unfortunately.